File: iafactory-gateway-python/providers/gemini.py
# -*- coding: utf-8 -*-
import httpx
import os

class GeminiProvider:
    def __init__(self):
        self.api_key = os.getenv("GEMINI_API_KEY", "")

    async def chat_completion(
        self,
        model: str,
        messages: list,
        temperature: float = 0.7,
        max_tokens: int = 1000,
        stream: bool = False,
    ):
        contents = []
        for m in messages:
            role = "user" if m.get("role") in ["user", "system"] else "model"
            contents.append({
                "role": role,
                "parts": [{"text": m.get("content", "")}],
            })

        payload = {
            "contents": contents,
            "generationConfig": {
                "temperature": temperature,
                "maxOutputTokens": max_tokens,
            },
        }

        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                f"https://generativelanguage.googleapis.com/v1/models/{model}:generateContent",
                params={"key": self.api_key},
                json=payload,
            )
            response.raise_for_status()
            data = response.json()

            return {
                "choices": [{
                    "message": {
                        "role": "assistant",
                        "content": data["candidates"][0]["content"]["parts"][0]["text"],
                    },
                    "finish_reason": "stop",
                }],
                "usage": data.get("usageMetadata", {}),
            }

gemini_provider = GeminiProvider()

File: apps/dzir-ia/backend/src/services/llm/gemini.ts
import type { ChatMsg } from "./openai_compat.js";

export async function chatGemini(opts: {
    apiKey: string;
    // ex: "gemini-2.0-flash"
    messages: ChatMsg[];
    model: string;
}) {
    const { apiKey, model, messages } = opts;

    // systemInstruction (Gemini)
    const systemText = messages.filter(m => m.role === "system").map(m => m.content).join("\n").trim();
    const contents = messages
        .filter(m => m.role !== "system")
        .map(m => ({
            parts: [{ text: m.content }],
            role: m.role === "assistant" ? "model" : "user",
        }));

    const url = `https://generativelanguage.googleapis.com/v1beta/models/${encodeURIComponent(
        model
    )}:generateContent?key=${encodeURIComponent(apiKey)}`;

    const body: any = { contents };
    if (systemText) {
        body.systemInstruction = { parts: [{ text: systemText }] };
    }

    const res = await fetch(url, {
        body: JSON.stringify(body),
        headers: { "content-type": "application/json" },
        method: "POST",
    });

    if (!res.ok) {
        const text = await res.text().catch(() => "");
        throw new Error(`Gemini ${res.status}: ${text}`);
    }

    const data: any = await res.json();
    const text =
        data?.candidates?.[0]?.content?.parts?.map((p: any) => p.text).filter(Boolean).join("") ?? "";
    return { raw: data, text };
}
